### PR TITLE
yaml: Generic yaml file parsing to configure device and link

### PR DESCRIPTION
**Pull request by ntsbtz**

### CODE REVIEW

File: src/manager/network-manager.c [hunk 1]
```diff
@@ -1930,28 +1929,45 @@ int manager_parse_proxy_config(GHashTable **c) {
...
-        r = generate_network_config(n);
-        if (r < 0) {
-                        log_warning("Failed to generate network configuration for file '%s': %s", file, strerror(-r));
+        if (l->ifname) {
+                r = parse_ifname_or_index(l->ifname, &p);
+                if (r < 0) {
+                        log_warning("Failed to find link '%s': %s", n->ifname, g_strerror(-r));
                         return r;
-        }
+                }
```
Comment (ssahani) [inline src/manager/network-manager.c:1950]: extra space

File: src/yaml/yaml-network-parser.c [hunk 2]
```diff
@@ -260,23 +296,27 @@ static int parse_yaml_node(YAMLManager *m, yaml_document_t *dp, yaml_node_t *nod
...
-
                 for (p = node->data.mapping.pairs.start; p < node->data.mapping.pairs.top; p++) {
                         n = yaml_document_get_node(dp, p->key);
-                        if (n)
-                                (void) parse_yaml_node(m, dp, n, network);
-
-                        n = yaml_document_get_node(dp, p->value);
-                        if (n)
-                                (void) parse_yaml_node(m, dp, n, network);
+                        if (string_equal(scalar(n), "match") || string_equal(scalar(n), "link")) {
```
Comment (ssahani) [inline src/yaml/yaml-network-parser.c:306]: `match` is for https://www.freedesktop.org/software/systemd/man/systemd.link.html#%5BMatch%5D%20Section%20Options. Only ```link:``` is useful

**Review by ssahani — CHANGES_REQUESTED** [review]
Overall looks OK. I have to test it

File: src/yaml/yaml-network-parser.c (hunk 2 above)
Comment (ntsbtz) [reply src/yaml/yaml-network-parser.c]: done

File: src/manager/network-manager.c [hunk 3]
```diff
@@ -1930,28 +1929,45 @@ int manager_parse_proxy_config(GHashTable **c) {
...
         _cleanup_(g_string_unrefp) GString *wifi_config = NULL;
         _cleanup_(network_unrefp) Network *n = NULL;
+        _cleanup_(netdev_link_unrefp) NetDevLink *l = NULL;
+        _auto_cleanup_ IfNameIndex *p = NULL;
         int r;
 
         assert(file);
 
-        r = parse_yaml_network_file(file, &n);
+
```
Comment (ssahani) [inline src/manager/network-manager.c:1938]: Extra line

File: src/manager/network-manager.c [hunk 4]
```diff
@@ -1930,28 +1929,45 @@ int manager_parse_proxy_config(GHashTable **c) {
 int manager_generate_network_config_from_yaml(const char *file) {
         _cleanup_(g_string_unrefp) GString *wifi_config = NULL;
         _cleanup_(network_unrefp) Network *n = NULL;
+        _cleanup_(netdev_link_unrefp) NetDevLink *l = NULL;
```
Comment (ssahani) [inline src/manager/network-manager.c:1932]: Variables should be reverse tree order

File: src/yaml/yaml-network-parser.c [hunk 5]
```diff
@@ -260,23 +296,27 @@ static int parse_yaml_node(YAMLManager *m, yaml_document_t *dp, yaml_node_t *nod
...
-                                (void) parse_yaml_node(m, dp, n, network);
+                        if (string_equal(scalar(n), "link")) {
+                                n = yaml_document_get_node(dp, p->value);
+                                if (n)
+                                        (void) parse_link_config(m, dp, n, link);
+                        } else if (string_equal(scalar(n), "device")) {
+                                n = yaml_document_get_node(dp, p->value);
+                                if (n) {
+                                        (void) parse_network_config(m, dp, n, network);
+                                }
```
Comment (ssahani) [inline src/yaml/yaml-network-parser.c:314]: extra `{}`

File: src/yaml/yaml-network-parser.c [hunk 6]
```diff
@@ -334,16 +383,17 @@ int parse_yaml_network_file(const char *file, Network **ret) {
...
         }
 
         yaml_parser_delete(&parser);
 
-        if (r >= 0)
-                *ret = steal_pointer(network);
-
+        if (r >= 0) {
+                *n = steal_pointer(network);
+                *l = steal_pointer(link);
```
Comment (ssahani) [inline src/yaml/yaml-network-parser.c:395]: hmm either this will be a network or link. Both can be same together. We need to take care of it

**Review by ssahani — CHANGES_REQUESTED** [review]

**Review by ssahani — APPROVED** [review]

**Pull request merged**